### PR TITLE
Initial port of ImperasDV to 40S

### PR DESCRIPTION
**Pull request by silabs-hfegran**
This commit aims to bring 40s/dev to the same state as 40x wrt. imperasDV

### CODE REVIEW

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert.sv [hunk 1]
```diff
@@ -824,20 +824,20 @@ module uvmt_cv32e40s_xsecure_assert
     ) else `uvm_error(info_tag, "mintstatus cs and shadow register missmatch does not result in major alert.\n");
 
     //MINTTHRESH:
-    a_xsecure_hardened_CSRs_missmatch_mintthresh: assert property (
-      p_xsecure_hardned_csr_missmatch_sets_alert_major(
-        xsecure_if.dut_wrap_cv32e40s_wrapper_i_core_i_cs_registers_i_smclic_csrs_mintthresh_csr_i_rdata_q,
-        xsecure_if.dut_wrap_cv32e40s_wrapper_core_cs_registers_smclic_csrs_mintthresh_csr_gen_hardened_shadow_q,
-        cv32e40s_pkg::CSR_MINTTHRESH_MASK)
-    ) else `uvm_error(info_tag, "mintthresh cs and shadow register missmatch does not result in major alert.\n");
+    //a_xsecure_hardened_CSRs_missmatch_mintthresh: assert property (
```
Comment (silabs-mateilga) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert.sv:827]: Is it intentional to merge this temporary disabling?

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert.sv [hunk 2]
```diff
@@ -664,10 +664,10 @@ module uvmt_cv32e40s_xsecure_assert
         and xsecure_if.dut_wrap_cv32e40s_wrapper_core_cs_registers_smclic_csrs_mintstatus_csr_gen_hardened_shadow_q == ~(xsecure_if.dut_wrap_cv32e40s_wrapper_i_core_i_cs_registers_i_smclic_csrs_mintstatus_csr_i_rdata_q & cv32e40s_pkg::CSR_MINTSTATUS_MASK)
 
         //MINTTHRESH
-        and xsecure_if.dut_wrap_cv32e40s_wrapper_core_cs_registers_smclic_csrs_mintthresh_csr_gen_hardened_shadow_q == ~(xsecure_if.dut_wrap_cv32e40s_wrapper_i_core_i_cs_registers_i_smclic_csrs_mintthresh_csr_i_rdata_q & cv32e40s_pkg::CSR_MINTTHRESH_MASK)
```
Comment (silabs-hfegran) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_xsecure_assert.sv:667]: Note, fixes to these assertions are inbound, commented out for now as this code fails to compile

**Comment by silabs-robin** [comment]
Before hitting merge, make sure that this scenario doesn't happen: 1) xsecure asserts fixes PR merged, 2) this PR overwrites with commented-out code.

**Comment by silabs-hfegran** [comment]
> Before hitting merge, make sure that this scenario doesn't happen: 1) xsecure asserts fixes PR merged, 2) this PR overwrites with commented-out code.

This should be resolved now

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv [hunk 3]
```diff
@@ -1325,6 +1337,13 @@ generate for (genvar n = 0; n < uvmt_cv32e40s_pkg::CORE_PARAM_PMP_NUM_REGIONS; n
    longint start_ovpsim_init_time;
    longint end_ovpsim_init_time;
    `ifndef FORMAL // Formal ignores initial blocks, avoids unnecessary warning
+
+   // overcome race
+   initial begin
```
Comment (MikeOpenHWGroup) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv:1342]: Can we move this initial block into the test entry point?   There should only be _one_ initial block in the entire environment.  (No, that is not (yet) in our coding style guidelines.)

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv [hunk 4]
```diff
@@ -1325,6 +1337,13 @@ generate for (genvar n = 0; n < uvmt_cv32e40s_pkg::CORE_PARAM_PMP_NUM_REGIONS; n
...
+      #0.9ns;
```
Comment (MikeOpenHWGroup) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv:1343]: Why is this here?  Why is it 0.9ns?

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv [hunk 5]
```diff
@@ -1397,7 +1416,16 @@ generate for (genvar n = 0; n < uvmt_cv32e40s_pkg::CORE_PARAM_PMP_NUM_REGIONS; n
 
       void'(uvm_config_db#(bit)::get(null, "", "sim_finished", sim_finished));
 
-      $display("\n%m: *** Test Summary ***\n");
+      // FIXME
```
Comment (MikeOpenHWGroup) [inline cv32e40s/tb/uvmt/uvmt_cv32e40s_tb.sv:1419]: Let's fix it now.  We do not control the RM via a cfg object, but we do control it via plusargs.  Recommend we do the same here.

File: cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_0/riscv_arithmetic_basic_test_0.S [hunk 6]
```diff
@@ -11961,7 +11961,7 @@ mmode_intr_vector_15:
                   j mmode_intr_handler
                   1: j test_done
 
-.align 2
+.align 7
```
Comment (MikeOpenHWGroup) [inline cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_0/riscv_arithmetic_basic_test_0.S:11964]: 7?!?   This is outside of my wheelhouse,  but why 7?

File: cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_1/riscv_arithmetic_basic_test_1.S [hunk 7]
```diff
@@ -12007,7 +12007,7 @@ mmode_intr_vector_15:
                   j mmode_intr_handler
                   1: j test_done
 
-.align 2
+.align 7
```
Comment (MikeOpenHWGroup) [inline cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_1/riscv_arithmetic_basic_test_1.S:12010]: Same question about 7.

File: lib/uvm_agents/uvma_rvvi/uvma_rvvi_tdefs.sv [hunk 8]
```diff
@@ -31,15 +31,20 @@ typedef enum int unsigned {
    UVMA_RVVI_HALT  = 2
 } uvma_rvvi_rm_action;
 
+typedef enum int unsigned {
+   UVMA_RVVI_V1 = 0,
+   UVMA_RVVI_V2 = 1
+} uvma_rvvi_version_enum;
```
Comment (MikeOpenHWGroup) [inline lib/uvm_agents/uvma_rvvi/uvma_rvvi_tdefs.sv:37]: I was unaware (or forgot) that there are multiple versions of RVVI released.  (In GitHub it's v1.6 only.)

File: mk/Common.mk [hunk 9]
```diff
@@ -182,14 +182,19 @@ endif
...
 
 DV_OVPM_HOME    = $(CORE_V_VERIF)/vendor_lib/imperas
 DV_OVPM_MODEL   = $(DV_OVPM_HOME)/imperas_DV_COREV
 DV_OVPM_DESIGN  = $(DV_OVPM_HOME)/design
 OVP_MODEL_DPI   = $(DV_OVPM_MODEL)/bin/Linux64/imperas_CV32.dpi.so
 #OVP_CTRL_FILE   = $(DV_OVPM_DESIGN)/riscv_CV32E40P.ic
 
+###############################################################################
+# Imperas OVPsim Instruction Set Simulator
+#IMPERAS_DV_MODEL = $(CORE_V_VERIF)/vendor_lib/ImperasDV/lib/Linux64/ImperasLib/imperas.com/verification/riscv/1.0/model.so
```
Comment (MikeOpenHWGroup) [inline mk/Common.mk:195]: Can we delete this?

File: mk/uvmt/dsim.mk [hunk 10]
```diff
@@ -35,8 +35,7 @@ DSIM_CODE_COV_SCOPE    ?= $(MAKE_PATH)/../tools/dsim/ccov_scopes.txt
 DSIM_USE_ISS           ?= YES
 
 DSIM_FILE_LIST         ?= -f $(DV_UVMT_PATH)/uvmt_$(CV_CORE_LC).flist
-DSIM_FILE_LIST         += -f $(DV_UVMT_PATH)/imperas_iss.flist
-DSIM_COMPILE_ARGS      += +define+$(CV_CORE_UC)_TRACE_EXECUTION
+DSIM_COMPILE_ARGS      += +define+$(CV_CORE_UC)_TRACE_EXECUTION+UVM
```
Comment (MikeOpenHWGroup) [inline mk/uvmt/dsim.mk:38]: Thank you very much for doing this.

**Review by MikeOpenHWGroup — CHANGES_REQUESTED** [review]
Brilliant!  Good job with this @silabs-hfegran.  I have a lot of comments, but few are "gating".  Mostly I just want to ensure we discuss these before we merge this into openhw/core-v-verif, to ensure we are aligned on things.

File: cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_0/riscv_arithmetic_basic_test_0.S (hunk 6 above)
Comment (silabs-hfegran) [reply cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_0/riscv_arithmetic_basic_test_0.S:11964]: This one puzzled me as I also initially thought this was byte alignment - It appears that risc-v is using size-log-2 alignment. log2(128) = 7: 

From
image

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_imperas_dv_wrap.sv [hunk 11]
```diff
@@ -0,0 +1,768 @@
...
+   generate
+       for(gi=0; gi<32; gi++)
+           assign rvvi.x_wdata[0][0][gi] = XREG[gi];
+   endgenerate
+
+   always @(*) begin
+       int i;
+       for (i=1; i<32; i++) begin
+           XREG[i] = 32'b0;
+           // TODO: This current RVFI implementation will only allow a single destination
```
[the remark this answers is not in the document]
Comment (MikeOpenHWGroup) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_imperas_dv_wrap.sv:407]: Push and Pop instructions are able to update multiple CSRs in a single instruction.

File: cv32e40s/tb/uvmt/uvmt_cv32e40s_iss_wrap.sv [hunk 12]
```diff
@@ -45,7 +45,7 @@ module uvmt_cv32e40s_iss_wrap
                 .ROM_BYTE_SIZE(ROM_BYTE_SIZE),
                 .RAM_BYTE_SIZE(RAM_BYTE_SIZE)) ram(bus);
 
-   CPU #(.ID(ID), .VARIANT("CV32E40S")) cpu(bus, io);
+   CPU #(.ID(ID), .VARIANT("CV32E40S_DEV")) cpu(bus, io);
```
[the remark this answers is not in the document]
Comment (MikeOpenHWGroup) [reply cv32e40s/tb/uvmt/uvmt_cv32e40s_iss_wrap.sv:48]: SiLabs asked for a 2nd model.  _May_ be deprecated in the future.

File: cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_1/riscv_arithmetic_basic_test_1.S (hunk 7 above)
Comment (MikeOpenHWGroup) [reply cv32e40s/tests/programs/custom/riscv_arithmetic_basic_test_1/riscv_arithmetic_basic_test_1.S:12010]: Add a comment to explain.

**Review by MikeOpenHWGroup — APPROVED** [review]
